diff --git a/src/dxvk/dxvk_graphics.cpp b/src/dxvk/dxvk_graphics.cpp
index 2c5334d..630dff6 100644
--- a/src/dxvk/dxvk_graphics.cpp
+++ b/src/dxvk/dxvk_graphics.cpp
@@ -176,7 +176,7 @@ namespace dxvk {
         && a.offset     == b.offset;
     }
 
-    for (uint32_t i = 0; i < viDivisorInfo.vertexBindingDivisorCount; i++) {
+    for (uint32_t i = 0; i < viDivisorInfo.vertexBindingDivisorCount && eq; i++) {
       const auto& a = viDivisors[i];
       const auto& b = other.viDivisors[i];
 
